lib/veri.ml

```diff
@@ -0,0 +1,207 @@
+open Core_kernel.Std
+open Bap.Std
+open Bap_traces.Std
+module Bap_value = Value
+open Bil.Result
+open Monad.State
+open Trace
+
+let do_nothing () = return ()
+
+type point = {
+  code : Chunk.t;
+  side : event list
+}
+let is_first_read ctxt ev =
+  match Bap_value.get Event.register_read ev with
+  | None -> false
+  | Some mv -> Option.is_none (ctxt#lookup (Move.cell mv))
+
+let is_first_load ctxt ev = 
+  match Bap_value.get Event.memory_load ev with
+  | None -> false
+  | Some mv ->
+    match Seq.find ~f:(fun (v,_) -> v = ctxt#mem) ctxt#bindings with
+    | None -> true
+    | Some (_,s) ->
+      match value s with
+      | Bil.Mem storage -> Option.is_none (storage#load (Move.cell mv))
+      | _ -> true 
+
+let is_init_event ctxt ev =
```

Comment:
Why do you need all this mess with `init` events? You shouldn't touch them at all, what if we will add new events, how you will extend the property of being initial?